tests/test_conformer_rel_pos.py

```diff
@@ -0,0 +1,121 @@
+    assert get_output_shape(10, 10, 20, kernel_size=3) == (10, 10, 20)  # odd kernel size
+    assert get_output_shape(5, 480, 512, dropout_broadcast_axes="T") == (5, 480, 512)  # dropout broadcast to T
+    assert get_output_shape(5, 480, 512, dropout_broadcast_axes="BT") == (5, 480, 512)  # dropout broadcast to BT
+
+
+def test_ConformerPositionwiseFeedForwardV2():
+    def get_output_shape(input_shape, input_dim, hidden_dim, dropout, activation, dropout_broadcast_axes=None):
+        x = torch.randn(input_shape)
+        cfg = ConformerPositionwiseFeedForwardV2Config(
+            input_dim=input_dim,
+            hidden_dim=hidden_dim,
+            dropout=dropout,
+            activation=activation,
+            dropout_broadcast_axes=dropout_broadcast_axes,
+        )
+        conf_ffn_part = ConformerPositionwiseFeedForwardV2(cfg)
+        y = conf_ffn_part(x)
+        return y.shape
+
+    for input_dim, hidden_dim, dropout, activation, dropout_broadcast_axes in product(
+        [10, 20], [100, 200], [0.1, 0.3], [nn.functional.silu, nn.functional.relu], [None, "B", "T", "BT"]
+    ):
+        input_shape = (10, 100, input_dim)
+        assert get_output_shape(input_shape, input_dim, hidden_dim, dropout, activation) == input_shape
+
+
+def test_ConformerMHSARelPosV1():
+    def get_output_shape(
+        input_shape,
+        seq_len,
+        input_dim,
+        num_att_heads=8,
+        att_weights_dropout=0.1,
+        dropout=0.1,
+        learnable_pos_emb=True,
+        rel_pos_clip=16,
+        with_pos_bias=False,
+        pos_emb_dropout=0.0,
+        dropout_broadcast_axes=None,
+    ):
+        assert len(input_shape) == 3 and input_shape[-1] == input_dim
+
+        cfg = ConformerMHSARelPosV1Config(
+            input_dim=input_dim,
+            num_att_heads=num_att_heads,
+            att_weights_dropout=att_weights_dropout,
+            dropout=dropout,
+            learnable_pos_emb=learnable_pos_emb,
+            rel_pos_clip=rel_pos_clip,
+            with_pos_bias=with_pos_bias,
+            pos_emb_dropout=pos_emb_dropout,
+            dropout_broadcast_axes=dropout_broadcast_axes,
+        )
+        conf_mhsa_rel_pos = ConformerMHSARelPosV1(cfg)
+        input_tensor = torch.randn(input_shape)
+        sequence_mask = torch.less(torch.arange(input_shape[1])[None, :], torch.tensor(seq_len)[:, None])
+
+        output = conf_mhsa_rel_pos(input_tensor, sequence_mask)
+
+        return list(output.shape)
+
+    # with key padding mask
+    input_shape = [4, 15, 32]  # B,T,F
+    seq_len = [15, 12, 10, 15]
+
+    for learnable_pos_emb, with_pos_bias, pos_emb_dropout in product([True, False], [True, False], [0.0, 0.1]):
+        assert get_output_shape(
+            input_shape,
+            seq_len,
+            32,
+            learnable_pos_emb=learnable_pos_emb,
+            with_pos_bias=with_pos_bias,
+            pos_emb_dropout=pos_emb_dropout,
+        ) == [4, 15, 32]
```

Comment:
I think I would add some test case which compares the output to ESPnet, so that you are sure to have it 100% correct.

Reply:
I did a local comparison, the outputs were close enough. but I needed to modify our `ConformerMHSARelPosV1` to make the parameters compatible, e.g. espnet uses separate q, k and v matrices.

Reply:
You don't need to modify this. When you import the model parameters, you just need to convert it properly. I think it makes sense to add such a test case.

Reply:
In the end I did find something wrong in my code by comparing.

Reply:
Yea, that's what I mean, you always find differences. It's always good to do such exact test.